central/reports/scheduler/v2/schedule.go

```diff
@@ -0,0 +1,646 @@
+		s.cron.Remove(oldEntryID)
+		delete(s.reportConfigToEntryIDs, reportConfigID)
+	}
+}
+
+func (s *scheduler) SubmitReport(request *ReportRequest, reSubmission bool) (string, error) {
+	request.Ctx = selectContext(request)
+	var err error
+	if request.ReportMetadata.GetReportStatus().GetReportRequestType() == storage.ReportStatus_ON_DEMAND {
+		userHasAnotherReport, err := s.doesUserHavePendingReport(request.Ctx, request.ReportConfig.GetId(), request.ReportMetadata.GetRequester().GetId())
+		if err != nil {
+			return "", err
+		}
+		if userHasAnotherReport {
+			return "", errors.Errorf("User already has a running report for config ID %s", request.ReportConfig.GetId())
+		}
+	}
+
+	request.ReportMetadata.ReportStatus.RunState = storage.ReportStatus_WAITING
+	request.ReportMetadata.ReportStatus.QueuedAt = types.TimestampNow()
+	if !reSubmission {
+		request.ReportMetadata.ReportId, err = s.reportMetadataStore.AddReportMetadata(request.Ctx, request.ReportMetadata)
+	} else {
+		err = s.reportMetadataStore.UpdateReportMetadata(request.Ctx, request.ReportMetadata)
+	}
+
+	if err != nil {
+		return "", err
+	}
+
+	s.lock.Lock()
+	s.queuedReports = append(s.queuedReports, request)
+	s.lock.Unlock()
+	s.sendNewReports <- true
+
+	return request.ReportMetadata.GetReportId(), nil
+}
+
+func (s *scheduler) doesUserHavePendingReport(ctx context.Context, configID string, userID string) (bool, error) {
+	query := search.NewQueryBuilder().
+		AddExactMatches(search.ReportConfigID, configID).
+		AddExactMatches(search.ReportState, storage.ReportStatus_WAITING.String(), storage.ReportStatus_PREPARING.String()).
+		AddExactMatches(search.ReportRequestType, storage.ReportStatus_ON_DEMAND.String()).
+		ProtoQuery()
+	runningReports, err := s.reportMetadataStore.SearchReportMetadatas(ctx, query)
+	if err != nil {
+		return false, err
+	}
+	for _, rep := range runningReports {
+		if rep.GetRequester().GetId() == userID {
+			return true, nil
+		}
+	}
+	return false, nil
+}
+
+func (s *scheduler) CancelReport(reportID string) {
+	//TODO implement me
+	panic("implement me")
+}
+
+func (s *scheduler) runReports() {
+	defer s.stopper.Flow().ReportStopped()
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case req := <-s.reportsToRun:
+			log.Infof("Executing report '%s' at %v", req.ReportConfig.GetName(), time.Now().Format(time.RFC822))
+			go s.sendReportResults(req)
+		}
+	}
+}
+
+func (s *scheduler) scheduleReports() {
+	maxConcurrency := int32(env.ReportExecutionMaxConcurrency.IntegerSetting())
+	for {
+		select {
+		case <-s.stopper.Flow().StopRequested():
+			return
+		case <-s.sendNewReports:
+			numRoutines := s.numRoutines.Load()
+			// Checking len(s.reportsToRun) < maxConcurrency (which is also the capacity of buffered channel reportsToRun) here
+			// will ensure that we never attempt to write to full channel reportsToRun. This will further ensure that
+			// writes to reportsToRun in selectRunnableReports() are never blocked.
```

Comment:
This is pretty complicated and it's hard to reason about the correctness.

Couple of ideas (though I haven't fully vetted it):

* Spawn N goroutines at the start where N == max concurrency. These goroutines will run `runReports` and just pick up jobs from `reportsToRun`. If there's something in the channel, it'll run otherwise it'll wait. At this point you don't even really need to buffer the chan (or keep it high) since there are only N goroutines so no more than N reports get run at any time. 
   *  Make sure you use a waitgroup and stop all of those from the `Stop` of the scheduler as well.
* Use [WeightedSemaphore](https://pkg.go.dev/golang.org/x/sync/semaphore). An example should be in pruning. Basically you create one set to the max concurrency and then TryAcquire(1) in this method. The individual `sendReportResults` goroutines will Release(1).

Reply:
Yeah the whole two long running routines approach for scheduling reports was pretty complicated. Updated to single routine using weightedSemaphore approach similar to the one used in `central/pruning/pruning.go` by `removeOrphanedNetworkFlows` func. Now `runReports` will be the only long running routine which will spawn a short routine for each report.